ssd1306 device driver example

DevyatovAndrey opened the pull request:
This adds step by step example of writing kernel device driver for OLED LCD display based on ssd1306 controller.

AleksandrBulyshchenko commented on ssd1306/Makefile:
```diff
@@ -0,0 +1,22 @@
+obj-m := ssd1306.o
+
+KDIR ?= ${HOME}/BBB/bb_kernel/
+#KDIR ?= ~/BBB/KERNEL/
```
Please avoid hardcodding any local path.
And unify expected env. vars with existent mpu6050 driver (see [its makefile](https://github.com/Kernel-GL-HRK/gl-kernel-training-2017/blob/master/mpu6050/Makefile))

DevyatovAndrey replied:
done

AleksandrBulyshchenko commented on ssd1306/build_on_x86.sh:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash -e
+
+export CROSS_COMPILE=${HOME}/BBB/gcc-linaro-5.3.1-2016.05-x86_64_arm-linux-gnueabihf/bin/arm-linux-gnueabihf-
```
This hardcodded export will override anything set in user's environment.
Please avoid committing your paths.

DevyatovAndrey replied:
done

AleksandrBulyshchenko commented on ssd1306/am335x-bone-common.dtsi:
```diff
@@ -0,0 +1,415 @@
+/*
```
Could you use the same baseline version of am335x-bone-common.dtsi as mpu6050 uses? (see ee49cc6ce7b1)

AleksandrBulyshchenko commented on userapp_analog_clock/Makefile:
```diff
@@ -0,0 +1,10 @@
+
+clean:
+	rm fb_user
+
```
This makefile has very odd view.
Does it work at all?

I believe you meant something like the following
```
CROSS_COMPILE ?= arm-linux-gnueabihf-
CC = $(CROSS_COMPILE)gcc
CFLAGS = -Wall

.PHONY: all clean

all: analog_clock

analog_clock: analog_clock.c
	$(CC) $(CFLAGS) $^ -o $@

clean:
	rm analog_clock
```

AleksandrBulyshchenko commented on userapp_analog_clock/build_user_app.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash -e
+
+export CROSS_COMPILE=${HOME}/BBB/gcc-linaro-arm-linux-gnueabihf-4.9-2014.09_linux/bin/arm-linux-gnueabihf-
```
Please avoid using hardcoded absolute paths - use just filename and add its location to `$PATH`

AleksandrBulyshchenko commented on userapp_analog_clock/build_user_app.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash -e
```
Why do you have shell script and makefile same time.
I believe something one is enough (preferably Makefile if you don't have to use some complex shell scripting)

DevyatovAndrey replied:
It's just to have an unified way to build and deploy this application

AleksandrBulyshchenko replied:
So which one is correct?
Now they use different options.
- Either `make` should be called from the shell script and everything have to be defined in the makefile (as it done for `ssd1306`);
- or one of them should be removed to avoid confusion.

DevyatovAndrey commented:
@AleksandrBulyshchenko,

- The branch is rebased onto this repo tree.
- Fixed checkpatch issues 
- Removed unnesessary export of enviroment variabless

AleksandrBulyshchenko commented on ssd1306/ssd1306.c:
```diff
@@ -88,11 +87,9 @@ static struct fb_var_screeninfo ssd1307fb_var = {
-    u8      data[LCD_WIDTH * LCD_HEIGHT / 8];
-
+	u8      data[LCD_WIDTH * LCD_HEIGHT / 8];
 } ssd1306_data_array;
```
`ssd1306.c:90: WARNING: do not add new typedefs`

AleksandrBulyshchenko commented on ssd1306/ssd1306.c:
```diff
@@ -159,26 +162,28 @@ static void ssd1306_init_lcd(struct i2c_client *drv_client) {
+
+	if (unlikely(ret < 0)) {
 		printk(KERN_ERR "i2c_master_send() has returned ERROR %d\n", ret);
 		return ret;
```
Better to use `pr_err()`

AleksandrBulyshchenko commented on envsetup.sh:
```diff
@@ -4,15 +4,15 @@ export CROSS_COMPILE=${HOME}/BBB/gcc-linaro-5.3.1-2016.05-x86_64_arm-linux-gnuea
 
```
I don't think this `envsetup.sh` should be published at all.
It will be different for everybody (and don't have to be at all)

DevyatovAndrey replied:
How to set up enviroment variables needed for these examples if there is no "envsetup.sh"?

AleksandrBulyshchenko replied:
The idea of using environment variables - is exactly to allow to use own environment, which is different for every user (otherwise they should be defined in build scripts instead of passing from environment).
So storing environment settings in VCS doesn't make much sense, because they either will be changed by users and will create a noise in git status or won't be used.

AleksandrBulyshchenko commented on ssd1306/build_on_x86.sh:
```diff
@@ -0,0 +1,39 @@
+                echo "configure kernel"
+                make ARCH=arm config
+                ;;
+            
```
redundant space line

AleksandrBulyshchenko commented on ssd1306/Makefile:
```diff
@@ -0,0 +1,23 @@
+PWD := $(shell pwd)
+
+default:
+		  $(MAKE) ARCH=arm  -C $(KDIR) M=$(PWD) modules
```
Please unify indentation - use single tab

DevyatovAndrey commented:
I've updated this PR:
- removed mistakenly added build artefacts
- unified am335x-bone-common.dtsi base with mpu6050
